text/000-buildpacks-contribute-default-process-type.md

```diff
@@ -0,0 +1,138 @@
+command = "<command to run>"
+default = <boolean (default=false)>
+```
+
+Example:
+
+```toml
+[[processes]]
+type = "web"
+command = "bundle exec ruby app.rb"
+default = true
+
+[[processes]]
+type = "worker"
+command = "bundle exec ruby worker.rb"
+```
+
+Buildpacks may set `default = true` to indicate that the process type being defined should be the default process type for the app image. If not specified, no default designation will be assumed (to be consistent with buildpacks implementing earlier buildpack APIs).
+
+If a buildpack attempts to define two processes with `default = true` specified, the lifecycle will fail.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+- Buildpacks could stomp on each other, each trying to declare different default process types. This might be confusing.
+- Changing the way things work now could surprise some people.
+- One more thing for buildpack authors to know and care about.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+  - Keep things as they are.
+  - Earlier iterations of this PR.
+- Why is this proposal the best? Easier for end users! It simplifies the build process (need to provide one fewer flag, don't need to rebuild if process types are not known ahead of time).
+- What is the impact of not doing this? A moderate amount of frustration.
+
+# Prior Art
+[prior-art]: #prior-art
+
+Discuss prior art, both the good and bad.
+
+`pack build` has a `--default-process` flag. If this flag is not provided by the user, `pack` will provide `-process-type web` to the lifecycle. If `web` is not a valid process type, the lifecycle will warn and set the image entrypoint to `/cnb/lifecycle/launcher`, which will require the user to specify the desired process in the command when running the app image. This has the advantage of "just working" for web apps. However for non-web apps, the user must know what process types exist and pass the desired default explicitly when building the app, otherwise they will get an image that must be configured at runtime.
+
+The lifecycle if not provided any `-process-type` will set the default process type to the only process type (if there is only one), otherwise it will warn and set the image entrypoint to `/cnb/lifecycle/launcher` as described above. This has the advantage of "just working" for single-process apps. However this requires the lifecycle to be "opinionated" and may be surprising for some users. See [spec PR](https://github.com/buildpacks/spec/pull/137) for the proposed removal of this behavior which should go hand-in-hand with this RFC.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Should `pack` continue to pass `-process-type web`?
+  - It could inspect the buildpack API and only pass the flag if the API version is below 0.X.
+  - Builders could set an environment variable to communicate to `pack` that it doesn't need to pass -process-type to the lifecycle
```

Comment:
@jromero it would be good to get your thoughts here.

Reply:
Relevant [Slack conversation](https://buildpacks.slack.com/archives/CSFE49VGR/p1602691248066000)

Reply:
For 3, what would happen if the buildpack is setting a default process type of "worker" but pack continues to pass "web"?

Am I right in assuming that the end result image would A. warn if "web" process type doesn't exist but continue to set "worker" as process type or B. set "web" process type if it exists.

Reply:
Seems like this conversation is also relevant: https://github.com/buildpacks/rfcs/pull/110/files#r504925751

May be worth discussing more at length (yet again, I'm sure) during a WG meeting.

Reply:
IIRC, yesterday during working group we explored the idea of easing the migration from a platform's perspective by changing the behaviour of `-process-type`. It would now fail if the process type is not available as opposed to simply warning as it currently does.

By making this change in conjunction to making buildpacks own setting default process types platforms have a more specific piece of data to look at and determine whether they should be passing in a "default" process type.

Reply:
Added a section to describe the migration approaches discussed: https://github.com/buildpacks/rfcs/pull/110/files#diff-00a2fb620563b974f876628cc19e746a4a96606f9dea0fd58dc1bb13f010b4abR139-R160

Reply:
Migration plan covered here: https://github.com/buildpacks/rfcs/pull/110/files#diff-00a2fb620563b974f876628cc19e746a4a96606f9dea0fd58dc1bb13f010b4abR103-R114